core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java

```diff
@@ -46,55 +48,90 @@
 @UnstableApi
 public abstract class AbstractRuleBuilder {
 
-    private final Predicate<RequestHeaders> requestHeadersFilter;
+    private final BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter;
 
     @Nullable
-    private Predicate<ResponseHeaders> responseHeadersFilter;
+    private BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter;
     @Nullable
-    private Predicate<HttpHeaders> responseTrailersFilter;
+    private BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter;
     @Nullable
-    private Predicate<Throwable> exceptionFilter;
+    private BiPredicate<ClientRequestContext, Throwable> exceptionFilter;
 
     /**
      * Creates a new instance with the specified {@code requestHeadersFilter}.
      */
     @SuppressWarnings("unchecked")
+    protected AbstractRuleBuilder(
+            BiPredicate<? super ClientRequestContext, ? super RequestHeaders> requestHeadersFilter) {
+        this.requestHeadersFilter = (BiPredicate<ClientRequestContext, RequestHeaders>) requestHeadersFilter;
+    }
+
+    /**
+     * Creates a new instance with the specified {@code requestHeadersFilter}.
+     *
+     * @deprecated Use {@link #AbstractRuleBuilder(BiPredicate)}.
+     */
+    @Deprecated
     protected AbstractRuleBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
-        this.requestHeadersFilter = (Predicate<RequestHeaders>) requestHeadersFilter;
+        this(toBiPredicateForSecond(requireNonNull(requestHeadersFilter, "requestHeadersFilter")));
```

Comment:
How about `BiPredicates.second()`?

Reply:
I thought that names too. I was not sure that the `second()` expresses the behavior. 🤔

- [Functions.second()](https://github.com/line/armeria/blob/ead4d1018db67aa68fe15e4ac237682f82662a61/core/src/main/java/com/linecorp/armeria/common/util/Functions.java#L207) returns a BiFunction that returns the second arguments.
- [BiPredicates.second(Predicate)] that returns a BiPredicate that uses only the second argument to evaluate the predicate.

How about `BiPredicates.ofSecond(Predicate)` at least to indicate the method is a factory?

Reply:
We're probably gonna remove it once the deprecated methods are gone, so.. maybe OK not to care about naming? 😅

Reply:
> We're probably gonna remove it once the deprecated methods are gone, so.. maybe OK not to care about naming? 😅

Yes, the `BiPredicates` class will be removed eventually.

Reply:
If so, then it's fine as it is. 😄